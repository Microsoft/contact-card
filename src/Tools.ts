import * as React from "react";
import { Button, BaseButton, LinkBase } from "office-ui-fabric-react";

export function openLink(url: string, e?: React.SyntheticEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement | HTMLDivElement | BaseButton | Button | LinkBase>) {
    let target: string | undefined = "_self";   // open in same window by default. This also prevents empty tabs for `mailto:` links in Chrome

    if (window.self !== window.top) {           // CSP (Content Security Policy) can prevent URL changing if we're inside iFrame. We let browser to decide where to open the URL
        target = undefined;
    }

    window.open(url, target);

    if (e) {
        e.preventDefault();
        e.stopPropagation();
    }
}
