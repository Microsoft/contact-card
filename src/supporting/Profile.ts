import { IPersonaProfile } from "../Types";


export function buildProfileResponse(id: number) {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        mail: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhones: [`555-123-456${id}`],
        imAddresses: [`im-user${id}@contoso.com`],
        companyName: "Contoso"
    };
}


export function buildProfile(id: number): IPersonaProfile {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        email: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhone: `555-123-456${id}`,
        imAddress: `im-user${id}@contoso.com`,
        companyName: "Contoso"
    };
}
