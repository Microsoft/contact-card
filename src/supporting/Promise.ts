export async function delay(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function waitNextTick() {
    return delay(0);
}
