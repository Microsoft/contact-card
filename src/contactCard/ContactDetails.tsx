import * as React from "react";
import { css, AnimationClassNames, Label, Link } from "office-ui-fabric-react";
import { IPersonaProfile } from "../Types";
import { openLink } from "../Tools";


export function renderContactDetails(profile: IPersonaProfile): React.ReactNode {
    return (
        <ul className={css("contact-details", AnimationClassNames.slideLeftIn400)}>
            <li>
                <Label>Email</Label>
                <Link href={`mailto:${profile.email}`} className="contact-link email" onClick={e => openLink(`mailto:${profile.email}`, e)}>{profile.email}</Link>
            </li>
            {profile.imAddress && (
                <li>
                    <Label>Chat</Label>
                    <Link href={`sip:${profile.imAddress}`} className="contact-link chat" onClick={e => openLink(`sip:${profile.imAddress}`, e)}>{profile.imAddress}</Link>
                </li>
            )}
            <li>
                <Label>Work phone</Label>
                <Link href={`tel:${profile.businessPhone}`} className="contact-link business-phone" onClick={e => openLink(`tel:${profile.businessPhone}`, e)}>{profile.businessPhone}</Link>
            </li>
            <li>
                <Label>Company</Label>
                <span>{profile.companyName}</span>
            </li>
            <li>
                <Label>Job title</Label>
                <span>{profile.jobTitle}</span>
            </li>
            <li>
                <Label>Department</Label>
                <span>{profile.department}</span>
            </li>
            <li>
                <Label>Office location</Label>
                <span>{profile.officeLocation}</span>
                <span>&nbsp;{profile.city}</span>
            </li>
        </ul>
    );
}
