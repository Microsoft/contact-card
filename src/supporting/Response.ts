import { buildProfileResponse } from "./Profile";

export function buildResponse(userId: number, status: number) {
    return new Response(JSON.stringify({
        responses: [
            {
                id: userId,
                body: buildProfileResponse(userId),
                status: status
            }
        ]
    }));
}
