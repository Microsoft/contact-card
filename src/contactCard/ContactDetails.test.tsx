import * as React from "react";
import * as Enzyme from "enzyme";
import { renderContactDetails } from "./ContactDetails";
import { buildProfile } from "../supporting/Profile";

jest.mock("../Tools");
import { openLink } from "../Tools";


test("renders full ContactDetails", () => {
    const wrap = Enzyme.shallow(<div>{renderContactDetails(buildProfile(1))}</div>);
    expect(wrap).toMatchSnapshot();
});


test("renders ContactDetails w/o IM", () => {
    const profile = buildProfile(1);
    profile.imAddress = undefined;
    const wrap = Enzyme.shallow(<div>{renderContactDetails(profile)}</div>);
    expect(wrap).toMatchSnapshot();
});


test("ContactDetails opens links", () => {
    const wrap = Enzyme.shallow(<div>{renderContactDetails(buildProfile(1))}</div>);
    wrap.find(".contact-link.email").simulate("click");
    wrap.find(".contact-link.chat").simulate("click");
    wrap.find(".contact-link.business-phone").simulate("click");
    expect(openLink).toBeCalledTimes(3);
});
