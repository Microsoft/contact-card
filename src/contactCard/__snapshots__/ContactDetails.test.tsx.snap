// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ContactDetails w/o IM 1`] = `
<div>
  <ul
    className="contact-details css-47"
  >
    <li>
      <StyledLabelBase>
        Email
      </StyledLabelBase>
      <StyledLinkBase
        className="contact-link email"
        href="mailto:user1@contoso.com"
        onClick={[Function]}
      >
        user1@contoso.com
      </StyledLinkBase>
    </li>
    <li>
      <StyledLabelBase>
        Work phone
      </StyledLabelBase>
      <StyledLinkBase
        className="contact-link business-phone"
        href="tel:555-123-4561"
        onClick={[Function]}
      >
        555-123-4561
      </StyledLinkBase>
    </li>
    <li>
      <StyledLabelBase>
        Company
      </StyledLabelBase>
      <span>
        Contoso
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Job title
      </StyledLabelBase>
      <span>
        Job Title 1
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Department
      </StyledLabelBase>
      <span>
        Department 1
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Office location
      </StyledLabelBase>
      <span>
        1234-1
      </span>
      <span>
         
        Redmond
      </span>
    </li>
  </ul>
</div>
`;

exports[`renders full ContactDetails 1`] = `
<div>
  <ul
    className="contact-details css-47"
  >
    <li>
      <StyledLabelBase>
        Email
      </StyledLabelBase>
      <StyledLinkBase
        className="contact-link email"
        href="mailto:user1@contoso.com"
        onClick={[Function]}
      >
        user1@contoso.com
      </StyledLinkBase>
    </li>
    <li>
      <StyledLabelBase>
        Chat
      </StyledLabelBase>
      <StyledLinkBase
        className="contact-link chat"
        href="sip:im-user1@contoso.com"
        onClick={[Function]}
      >
        im-user1@contoso.com
      </StyledLinkBase>
    </li>
    <li>
      <StyledLabelBase>
        Work phone
      </StyledLabelBase>
      <StyledLinkBase
        className="contact-link business-phone"
        href="tel:555-123-4561"
        onClick={[Function]}
      >
        555-123-4561
      </StyledLinkBase>
    </li>
    <li>
      <StyledLabelBase>
        Company
      </StyledLabelBase>
      <span>
        Contoso
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Job title
      </StyledLabelBase>
      <span>
        Job Title 1
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Department
      </StyledLabelBase>
      <span>
        Department 1
      </span>
    </li>
    <li>
      <StyledLabelBase>
        Office location
      </StyledLabelBase>
      <span>
        1234-1
      </span>
      <span>
         
        Redmond
      </span>
    </li>
  </ul>
</div>
`;
